src/platform/Linux/NetworkCommissioningWiFiDriver.cpp

```diff
@@ -123,96 +170,247 @@ void LinuxWiFiDriver::Shutdown()
+        !mStagingNetworks[mStagedConnectedNetworkIndex].Matches(
+            ByteSpan(configuredNetwork.networkID, configuredNetwork.networkIDLen)))
+    {
+        ConnectNetwork(
+            ByteSpan(mStagingNetworks[mStagedConnectedNetworkIndex].ssid, mStagingNetworks[mStagedConnectedNetworkIndex].ssidLen),
+            mpConnectCallback);
+    }
+
     return CHIP_NO_ERROR;
 }
 
 Status LinuxWiFiDriver::AddOrUpdateNetwork(ByteSpan ssid, ByteSpan credentials, MutableCharSpan & outDebugText,
                                            uint8_t & outNetworkIndex)
 {
     outDebugText.reduce_size(0);
-    outNetworkIndex = 0;
-    VerifyOrReturnError(mStagingNetwork.Empty() || mStagingNetwork.Matches(ssid), Status::kBoundsExceeded);
+    outNetworkIndex = mCurrentNetId;
+
+    VerifyOrReturnError(mCurrentNetId < kMaxNetworks, Status::kBoundsExceeded);
+    VerifyOrReturnError(mStagingNetworks[mCurrentNetId].Empty() || mStagingNetworks[mCurrentNetId].Matches(ssid),
+                        Status::kBoundsExceeded);
 
     // Do the check before setting the values, so the data is not updated on error.
-    VerifyOrReturnError(credentials.size() <= sizeof(mStagingNetwork.credentials), Status::kOutOfRange);
-    VerifyOrReturnError(!ssid.empty() && ssid.size() <= sizeof(mStagingNetwork.ssid), Status::kOutOfRange);
+    VerifyOrReturnError(credentials.size() <= sizeof(mStagingNetworks[mCurrentNetId].credentials), Status::kOutOfRange);
+    VerifyOrReturnError(!ssid.empty() && ssid.size() <= sizeof(mStagingNetworks[mCurrentNetId].ssid), Status::kOutOfRange);
 
-    memcpy(mStagingNetwork.credentials, credentials.data(), credentials.size());
-    mStagingNetwork.credentialsLen = static_cast<decltype(mStagingNetwork.credentialsLen)>(credentials.size());
+    memcpy(mStagingNetworks[mCurrentNetId].credentials, credentials.data(), credentials.size());
+    mStagingNetworks[mCurrentNetId].credentialsLen =
+        static_cast<decltype(mStagingNetworks[mCurrentNetId].credentialsLen)>(credentials.size());
 
-    memcpy(mStagingNetwork.ssid, ssid.data(), ssid.size());
-    mStagingNetwork.ssidLen = static_cast<decltype(mStagingNetwork.ssidLen)>(ssid.size());
+    memcpy(mStagingNetworks[mCurrentNetId].ssid, ssid.data(), ssid.size());
+    mStagingNetworks[mCurrentNetId].ssidLen = static_cast<decltype(mStagingNetworks[mCurrentNetId].ssidLen)>(ssid.size());
 
 #if CHIP_DEVICE_CONFIG_ENABLE_WIFI_PDC
-    mStagingNetwork.networkIdentityLen = 0;
-    mStagingNetwork.clientIdentityLen  = 0;
-    mStagingNetwork.clientIdentityKeypair.reset();
+    mStagingNetworks[mCurrentNetId].networkIdentityLen = 0;
+    mStagingNetworks[mCurrentNetId].clientIdentityLen  = 0;
+    mStagingNetworks[mCurrentNetId].clientIdentityKeypair.reset();
 #endif // CHIP_DEVICE_CONFIG_ENABLE_WIFI_PDC
+    mCurrentNetId++;
 
     return Status::kSuccess;
 }
 
+void LinuxWiFiDriver::ShiftNetworkAfterRemove()
```

Comment:
Maybe this should be called something like "CompressStagingNetworksList"?  Not sure.